lib/puppet/pops/types/type_mismatch_describer.rb

```diff
@@ -581,6 +581,28 @@ def validate_default_parameter(subject, param_name, param_type, value, tense = :
       end
     end
 
+    def get_deferred_function_return_type(value)
+      func = Puppet.lookup(:loaders).private_environment_loader.
+               load(:function, value.name)
+      dispatcher = func.class.dispatcher.find_matching_dispatcher(value.arguments)
+      raise ArgumentError, "No matching arity found for #{func.class.name} with arguments #{value.arguments}" unless dispatcher
+      dispatcher.type.return_type
+    end
+    private :get_deferred_function_return_type
+
+    def matching_return_type?(type, return_type)
+      if type.kind_of?(Puppet::Pops::Types::PVariantType)
+        if return_type.kind_of?(Puppet::Pops::Types::PVariantType)
+          matching_types = type.types & return_type.types
+          return !matching_types.empty?
+        end
+        type.types.include?(return_type)
+      else
+        type == return_type.class
+      end
+    end
+    private :matching_return_type?
```

Comment:
I think the type system can tell us if the deferred function's `return_type` is compatible with the parameter's `value_type` like:
```ruby
value_type.assignable?(df_return_type)
```
It seems to work for class parameters like: `Optional[String]`, `Variant[String, Integer]`.